### PR TITLE
code reviews

### PR DESCRIPTION
**Pull request by ps789**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,67 @@
...
+$ git clone 'https://github.com/openai/CLIP'
+```
+2. CLIP+VQGAN: Clone the required repositores and make edits:
+```bash
+$ cd CLIP+Optim
+$ git clone 'https://github.com/openai/CLIP'
+$ git clone 'https://github.com/CompVis/taming-transformers' tamingtransformers
+```
+
+Then we need to modify files within `tamingtransformers/taming` to have their
```
Comment (FelixHohne) [inline README.md:41]: Out of curiosity, why is this necessary?

File: CLIP+VQGAN/generate.py [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+if not torch.cuda.is_available():
+    default_image_size = 256  # no GPU found
+elif get_device_properties(0).total_memory <= 2 ** 33:  # 2 ** 33 = 8,589,934,592 bytes = 8 GB
+    default_image_size = 318  # <8GB VRAM
+
+# Create the parser
+vq_parser = argparse.ArgumentParser(description='Image generation using VQGAN+CLIP')
+
+# Add the arguments
+vq_parser.add_argument("-p",    "--prompts", type=str, help="Text prompts", default=None, dest='prompts')
```
Comment (FelixHohne) [inline CLIP+VQGAN/generate.py:46]: Consider moving your parser configuration into a separate file parse.py

File: CLIP+VQGAN/generate.py [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+        
+        if self.noise_fac:
+            facs = batch.new_empty([self.cutn, 1, 1, 1]).uniform_(0, self.noise_fac)
+            batch = batch + facs * torch.randn_like(batch)
+        return batch
+
+
+# An updated version with Kornia augments and pooling (where my version started):
+class MakeCutoutsPoolingUpdate(nn.Module):
+    def __init__(self, cut_size, cutn, args, cut_pow=1.):
```
Comment (FelixHohne) [inline CLIP+VQGAN/generate.py:159]: Modern python has pretty type hinting support. I don't know which classes are most exposed, but I generally find it helpful when reading code to know what type objects are. For example, it would be useful to know immediately if cut_size is an integer, float, etc.

File: CLIP+VQGAN/generate.py [hunk 4]
```diff
@@ -0,0 +1,649 @@
...
+    
+    with torch.no_grad():
+        z.copy_(z.maximum(z_min).minimum(z_max))
+
+
+def main():
+    # Execute the parse_args() method
+    args = vq_parser.parse_args()
+    if not args.prompts and not args.image_prompts:
+        args. prompts = "A cute, smiling, Nerdy Rodent"
```
Comment (FelixHohne) [inline CLIP+VQGAN/generate.py:334]: How does the dot space syntax work in Python?

**Review by FelixHohne — APPROVED** [review]

File: CLIP+VQGAN/generate.py [hunk 5]
```diff
@@ -0,0 +1,650 @@
...
+    
+    if args.make_video:    
+        img = np.array(out.mul(255).clamp(0, 255)[0].cpu().detach().numpy().astype(np.uint8))[:,:,:]
+        img = np.transpose(img, (1, 2, 0))
+        imageio.imwrite('./steps/' + str(i) + '.png', np.array(img))
+
+    return result # return loss
+
+
+def train(i, opt, z, model, perceptor, normalize, make_cutouts, args, pMs, z_min, z_max):
```
Comment (ReallyNeil) [inline CLIP+VQGAN/generate.py:315]: Similar comment to Felix, it would help a lot if type hinting was used. I find when systems become scalable type hinting becomes more and more valuable but ymmv

File: CLIP+VQGAN/generate.py [hunk 6]
```diff
@@ -0,0 +1,650 @@
...
+        return clamp_with_grad(torch.cat(cutouts, dim=0), 0, 1)
+
+
+def load_vqgan_model(config_path, checkpoint_path):
+    global gumbel
+    gumbel = False
+    config = OmegaConf.load(config_path)
+    if config.model.target == 'tamingtransformers.taming.models.vqgan.VQModel':
+        model = vqgan.VQModel(**config.model.params)
+        model.eval().requires_grad_(False)
```
Comment (ReallyNeil) [inline CLIP+VQGAN/generate.py:257]: Since this happens regardless of the case, maybe consider putting outside of if or elif statements

File: CLIP+VQGAN/generate.py [hunk 7]
```diff
@@ -0,0 +1,650 @@
...
+                            pil_image_zoom = zoom_at(pil_image, args.size[0]/2, args.size[1]/2, args.zoom_scale)
+                            
+                            # Convert image back to a tensor again
+                            pil_tensor = TF.to_tensor(pil_image_zoom)
+                            
+                            # Re-encode
+                            z, *_ = model.encode(pil_tensor.to(device).unsqueeze(0) * 2 - 1)
+                            z_orig = z.clone()
+                            z.requires_grad_(True)
+
```
Comment (ReallyNeil) [inline CLIP+VQGAN/generate.py:538]: Love that you comment every intermediate step of the method

File: CLIP+VQGAN/generate.py (hunk 4 above)
Comment (ReallyNeil) [reply CLIP+VQGAN/generate.py:334]: Personal Preference is to break main down into some smaller reference methods if it is thing large, but this may be the standard for ML code

File: CLIP+VQGAN/generate.py [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+    # Image initialisation
+    if args.init_image:
+        if 'http' in args.init_image:
+            img = Image.open(urlopen(args.init_image))
+        else:
+            img = Image.open(args.init_image)
+            pil_image = img.convert('RGB')
+            pil_image = pil_image.resize((sideX, sideY), Image.LANCZOS)
+            pil_tensor = TF.to_tensor(pil_image)
+            z, *_ = model.encode(pil_tensor.to(device).unsqueeze(0) * 2 - 1)
```
Comment (ReallyNeil) [inline CLIP+VQGAN/generate.py:425]: These four lines are repeated in all if/elif statements, consider minimizing code duplication

**Review by ReallyNeil — COMMENTED** [review]
Great Work! Learned a lot about project from combing through your work so far. Some minor concerns, mostly regarding code duplication, but overall very inspirational.
